src/algebra/jordan/triple.lean

```diff
@@ -0,0 +1,139 @@
+symmetric in the first and thrid variable and conjugate linear in the second variable. A *-triple
+satisfying the Leibniz rule is said to be a Jordan *-triple.
+
+As well as being of algebraic interest, Jordan *-triples arise naturally in mathematical physics,
+functional analysis and differential geometry. For more information about these connections the
+interested reader is referred to [alfsenshultz2003], [chu2012], [friedmanscarr2005],
+[iordanescu2003] and [upmeier1987].
+
+## Main results
+
+(None yet, we're just getting started.)
+
+## References
+
+* [Chu, Jordan Structures in Geometry and Analysis][chu2012]
+-/
+
+lemma biadd_expand {A : Type*} [add_comm_monoid A] (B: A →+ A →+ A) (a₁ a₂ b₁ b₂ : A) :
+  B (a₁ + a₂) (b₁+b₂) = B a₁ b₁ + B a₂ b₁ + B a₁ b₂ + B a₂ b₂ :=
+  by rw [map_add, map_add, add_monoid_hom.add_apply, add_monoid_hom.add_apply, ← add_assoc]
+
+/-- An additive commutative monoid with a trilinear triple product -/
+class has_trilinear_tp (A : Type*) [add_comm_monoid A] := (tp : A →+ A →+ A →+ A )
+
+notation ⦃a, b, c⦄ := has_trilinear_tp.tp a b c
+
+lemma add_left (A : Type*) [add_comm_monoid A] [has_trilinear_tp A] (a₁ a₂ b c : A) :
+  ⦃a₁ + a₂, b, c⦄ = ⦃a₁, b, c⦄ + ⦃a₂, b, c⦄ :=
+by rw [map_add, add_monoid_hom.add_apply, add_monoid_hom.add_apply]
+
+lemma add_middle (A : Type*) [add_comm_monoid A] [has_trilinear_tp A] (a b₁ b₂ c : A) :
+  ⦃a, b₁ + b₂, c⦄ = ⦃a, b₁, c⦄ + ⦃a, b₂, c⦄ := by rw [map_add, add_monoid_hom.add_apply]
+
+lemma add_right (A : Type*) [add_comm_monoid A] [has_trilinear_tp A] (a b c₁ c₂ : A) :
+  ⦃a, b, c₁ + c₂⦄ = ⦃a, b, c₁⦄ + ⦃a, b, c₂⦄ := by rw map_add
+
+section trilinear_product
+
+variables {A : Type*} [add_comm_monoid A] [has_trilinear_tp A]
+
+/-- Define the multiplication operator `D` -/
+def D : A →+ A →+ add_monoid.End A := has_trilinear_tp.tp
+
+/-- homotope a is the a-homotope -/
+def homotope : A →+ A →+ add_monoid.End A := (D : A →+ A →+ add_monoid.End A).flip_hom
+
+
+lemma homotope_def (a b c : A) : homotope b a c = ⦃a, b, c⦄ :=
+begin
+  unfold homotope,
+  unfold D,
+  rw [add_monoid_hom.flip_hom_apply, add_monoid_hom.flip_apply],
+end
+
+lemma lr_bilinear (a₁ a₂ b c₁ c₂ : A) : ⦃a₁ + a₂, b, c₁ + c₂⦄ =
+  ⦃a₁, b, c₁⦄ + ⦃a₂, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₂⦄ := by rw [← homotope_def,
+  biadd_expand (homotope b), homotope_def, homotope_def, homotope_def, homotope_def]
+
+end trilinear_product
+
+/-- A Jordan triple product satisfies a Leibniz law -/
+class is_jordan_tp (A : Type*) [add_comm_monoid A] [has_sub A] extends has_trilinear_tp A:=
+(comm : ∀ (a b c : A), ⦃a, b, c⦄ = ⦃c, b, a⦄)
+(leibniz : ∀ (a b c d e: A), ⦃a, b, ⦃c, d, e⦄⦄  =
+  ⦃⦃a, b, c⦄, d, e⦄ - ⦃c, ⦃b, a, d⦄, e⦄ + ⦃c, d, ⦃a, b, e⦄⦄)
+
+/--
+We say that a pair of operators $(T,T^′)$ are Leibniz if they satisfy a law reminiscent of
+differentiation.
+-/
+def leibniz {A : Type*} [add_comm_monoid A] [has_trilinear_tp A] (T : A → A) (T'  : A → A) :=
```

Comment:
Can you avoid picking very common names in the root namespace? I can think of a dozen other things that could be called `leibniz`.